types/priv_validator/socket.go

```diff
@@ -118,101 +182,142 @@ func (pvsc *PrivValidatorSocketClient) SignHeartbeat(chainID string, heartbeat *
 type PrivValidatorSocketServer struct {
 	cmn.BaseService
 
-	conn        net.Conn
-	proto, addr string
-	listener    net.Listener
+	proto, addr    string
+	listener       net.Listener
+	maxConnections int
+	privKey        *crypto.PrivKeyEd25519
 
 	privVal PrivValidator
 	chainID string
 }
 
-func NewPrivValidatorSocketServer(logger log.Logger, socketAddr, chainID string, privVal PrivValidator) *PrivValidatorSocketServer {
+// NewPrivValidatorSocketServer returns an instance of
+// PrivValidatorSocketServer.
+func NewPrivValidatorSocketServer(
+	logger log.Logger,
+	chainID, socketAddr string,
+	maxConnections int,
+	privVal PrivValidator,
+	privKey *crypto.PrivKeyEd25519,
+) *PrivValidatorSocketServer {
 	proto, addr := cmn.ProtocolAndAddress(socketAddr)
 	pvss := &PrivValidatorSocketServer{
-		proto:   proto,
-		addr:    addr,
-		privVal: privVal,
-		chainID: chainID,
+		proto:          proto,
+		addr:           addr,
+		maxConnections: maxConnections,
+		privKey:        privKey,
+		privVal:        privVal,
+		chainID:        chainID,
 	}
 	pvss.BaseService = *cmn.NewBaseService(logger, "privValidatorSocketServer", pvss)
 	return pvss
 }
 
+// OnStart implements cmn.Service.
 func (pvss *PrivValidatorSocketServer) OnStart() error {
-	if err := pvss.BaseService.OnStart(); err != nil {
-		return err
-	}
 	ln, err := net.Listen(pvss.proto, pvss.addr)
 	if err != nil {
 		return err
 	}
-	pvss.listener = ln
-	go pvss.acceptConnectionsRoutine()
+
+	pvss.listener = netutil.LimitListener(ln, pvss.maxConnections)
+
+	go pvss.acceptConnections()
+
 	return nil
 }
 
+// OnStop implements cmn.Service.
 func (pvss *PrivValidatorSocketServer) OnStop() {
-	pvss.BaseService.OnStop()
-	if err := pvss.listener.Close(); err != nil {
-		pvss.Logger.Error("Error closing listener", "err", err)
+	if pvss.listener == nil {
+		return
 	}
 
-	if err := pvss.conn.Close(); err != nil {
-		pvss.Logger.Error("Error closing connection", "conn", pvss.conn, "err", err)
+	if err := pvss.listener.Close(); err != nil {
+		pvss.Logger.Error("OnStop", "err", errors.Wrap(err, "closing listener failed"))
 	}
 }
 
-func (pvss *PrivValidatorSocketServer) acceptConnectionsRoutine() {
+func (pvss *PrivValidatorSocketServer) acceptConnections() {
 	for {
-		// Accept a connection
-		pvss.Logger.Info("Waiting for new connection...")
-		var err error
-		pvss.conn, err = pvss.listener.Accept()
+		conn, err := pvss.listener.Accept()
 		if err != nil {
 			if !pvss.IsRunning() {
 				return // Ignore error from listener closing.
 			}
-			pvss.Logger.Error("Failed to accept connection: " + err.Error())
+			pvss.Logger.Error(
+				"accpetConnections",
```

Comment:
"Failed to accept new connection"

Reply:
While I see the point given the signature of `Error(msg string, keyvals ...interface{})` it seems very hard to build up the context the error was coming from as it would result in: 

```
E[02-13|17:42:57.620] Failed to accept new connection                                      err=foo
```

Maybe I'm not seeing an important aspect of error logging in our codeabse, but that line has no signal in it. Happy to incorporate your comments, just wondering if it is the best way to go abouut it.

Reply:
You can turn on a tracing logger by providing `--trace` flag https://github.com/tendermint/tendermint/blob/master/cmd/tendermint/commands/root.go#L53

In this mode logger will output full trace if error supports this feature (pkg/errors does). https://github.com/tendermint/tmlibs/blob/master/log/tracing_logger.go#L58

Reply:
Thanks for pointing that out! This still doesn't really help after the fact, which makes errors emitted without tracing still not very useful.

Reply:
Ok. I see your point. So the agreement always was: `meaningful-message key=value`. For example: "Failed to accept connection err=EOF". We're still in a process of improving the messages. Plus we started to use `pkg/errors` not so long ago to accumulate context.

Usually, what we do is just grep by msg ("Failed to accept connection"). Of course, if you have many places in your app where you accept a connection, then msg must be changed to something like "Failed to accept connection from X".

I am happy to hear any suggestions on how to improve the current scheme.

Reply:
Very sound approach and I don't mean to start a greater discussion if we already iterating on error handling in other places.

> I am happy to hear any suggestions on how to improve the current scheme.

Usually thinking about in what remote ends logs can end up in typical infrastructure I try to imagine what vital info is needed to make sense of it when consumed in some log infrastructure remote from the origin. Additionally how each log line can interpreted by itself without context of the lines before or after. As the logger is passed from the outside to the inside in the program we can enrich the keyvals with information as well which subsystems don't need to care about. A standard set of fields I try to surface, most of them which can easily be achieved with Dynamic Contextual Values from the go-kit log package:

* **caller**: https://godoc.org/github.com/go-kit/kit/log#Caller
* **timestamp**: https://godoc.org/github.com/go-kit/kit/log#pkg-variables
* **revision**: compile flag: https://github.com/tendermint/tendermint/blob/master/Makefile#L10
* **job/system**: `tendermint` in our case
* **cmd/task**: `node`for the node command
* **service/subsystem**: what we pass to `cmn.NewBaseService(...)` e.g.: `privValidatorSocketServer`
* **err**: set if error present

Reply:
Very nice